lib/xapi-stdext-date/date.ml

```diff
@@ -37,19 +37,37 @@ type iso8601 = Ptime.date * Ptime.time * print_type
 let of_dt print_type dt = let (date, time) = dt in (date, time, print_type)
 let to_dt (date, time, _) = (date, time)
 
-let of_string x =
+let best_effort_iso8601_to_rfc3339 x =
+  (* (a) add dashes
+   * (b) add UTC tz if no tz provided *)
   let x =
     try
-      (* if x doesn't contain dashes, insert them, so that ptime can parse x *)
-      Scanf.sscanf x "%04d%02d%02dT%s" (fun y mon d rest ->
-        Printf.sprintf "%04d-%02d-%02dT%s" y mon d rest
-      )
-    with _ -> x
+      Scanf.sscanf x "%04d%02d%02dT%s"
+        (fun y mon d rest ->
+          Printf.sprintf "%04d-%02d-%02dT%s" y mon d rest)
+    with _ ->
+      x
+  in
+  let tz =
+    try
+      Scanf.sscanf x "%04d-%02d-%02dT%02d:%02d:%02d%s"
+        (fun _ _ _ _ _ _ tz -> Some tz)
+    with _ -> None
   in
-  match x |> Ptime.of_rfc3339 |> Ptime.rfc3339_error_to_msg with
-  | Error (`Msg e) -> invalid_arg (Printf.sprintf "date.ml:of_string: %s" e)
+  match tz with
+  | None | Some "" ->
+    (* the caller didn't specify a tz. we must try to add one so that ptime can at least attempt to parse *)
+    (Printf.sprintf "%sZ" x, PrintLocal)
+  | Some _ ->
+    (* the caller specified a tz. we assume it's UTC because we don't accept anything else *)
```

Comment:
I'm not too happy about ignoring the tz when it's not UTC. But not sure how to log this and it doesn't seem wise to fail here.

Reply:
I think adding a test with a different, valid timezone is something worth doing to provide evidence on the behaviour.

Reply:
There is already such a test (if a tz is specified it should be utc)

```
  let test_only_utc () =
    let utc = "2020-12-20T18:10:19Z" in
    let _ = of_string utc in (* UTC is valid *)
    let non_utc = "2020-12-20T18:10:19+02:00" in
    let exn = Invalid_argument "date.ml:of_string: 2020-12-20T18:10:19+02:00" in
    Alcotest.check_raises "only UTC is accepted" exn (fun () ->  of_string non_utc |> ignore)
  in
```